### DISCUSSION TITLE
Adding tools to help with clang-format to repo

### DISCUSSION THREAD

**Discussion by pmolodo** [Autodesk/maya-usd#1118] [Q&A]
Hi all - I wanted to add some utilties to the repo to make it easier to avoid clang-format issues.

At minimum, I wanted to add a binary command which will run clang-format on the entire repo (or specified files).

I'd like to also add support for [pre-commit](https://pre-commit.com/), to have an easy way to set up pre-commit hooks that ensure clang-format is run.  This would just entail adding an extra file, `.pre-commit-config.yaml`, to the repo root; for those that did not wish to use [pre-commit](https://pre-commit.com/), they could simply ignore it.  (Or, if they want to set up their own hooks, they could use it as a starting reference point.)

Would anyone else find this helpful, or have objections?

**Comment by kxl-adsk** [comment]
Hi Paul, All,

Developers have many different options available when setting up clang-format. Let's list them here:
- IDEs level integration - most commonly used have one, like Visual Studio or VS Code
- Git integration - using `git clang-format`
- Git hooks - using `pre-commit`
- Command-line executable - `clang-format -i`

I don't have any objections to the proposal. Just curious to find out which options people are preferring?

Cheers,
KxL

**Reply by ppt-adsk** [reply]
I usually use git clang-format, and clang-format -i on occasion.

**Reply by seando-adsk** [reply]
I also run "git clang-format" from the command-line and have it setup to auto-format in Visual Studio.

**Reply by pmolodo** [reply]
So, I've personally been using a combination of IDE integration and git hook (`pre-commit`).

I actually didn't realize that `git clang-format` was part of llvm / clang, though - so I might give that a shot.

I like the git hook, though, because it's a form of last-step-assurance... ie, it's easy to have non-formatted changes squeak in through other avenues (most common for me - when making edits in a merge-resolution tool), and this stops those before they're committed.  And, as I said, by simply providing the `.pre-commit-config.yaml` file, you're just giving folks the option for easily setting up `pre-commit` - for those who don't want it, it can be ignored.

Lastly - one thing none of these have (at least as far as I could tell?) was an easy way to run clang-format on the entire repo.  That's not necessarily something that needs to be done frequently, but can be handy on occasion.  Thus the addition of a separate [run-clang-format.py](https://github.com/Autodesk/maya-usd/blob/f437a577b5cde25036f93d1297090adadf564264/test/bin/run-clang-format.py) tool in [my PR](https://github.com/Autodesk/maya-usd/pull/1120).

Now that I know that `git clang-format` is available by default, though, if we decide we don't need / want a format-the-whole-repo tool, I can remove it and make the `.pre-commit-config.yaml` rely on `git clang-format`...
